### ISSUE TITLE
Getting validation errors when using `Swagger2->new($file)->expand->validate`

### ISSUE THREAD

**Issue by perlpunk** [jhthorsen/swagger2#80]
It complains about numbers when expecting booleans.
A test file will be added.
Like discussed on IRC.

**Comment by jhthorsen** [comment]
The issue is caused by the order of how the YAML file is loaded: I think the "coerce" information that JSON::Validator sets (based on the YAML input) is lost when JSON::Validator only sees the data structures - As pointed out by @perlpunk.

Thanks! 👍 

I might mark ```new($file)``` as deprecated if that's the easiest way to "untangle" the methods in Swagger2.

**Comment by dboehmer** [comment]
I wrote a short test file to check every possible combination:

```perl
use strict;
use warnings;

use Test::More;
use Swagger2;

my $file = "data://main/swagger2.yaml";
my $yaml = do { <DATA>; local $/; <DATA> };

sub t {
    my $perl   = shift;
    my @errors = eval $perl;
    is scalar(@errors) => 0, "$perl" or diag explain \@errors;
}

t 'Swagger2->new($file)->validate';
t 'Swagger2->new($file)->expand->validate';

t 'Swagger2->new($yaml)->validate';
t 'Swagger2->new($yaml)->expand->validate';

t 'Swagger2->new->load($file)->validate';
t 'Swagger2->new->load($file)->expand->validate';

t 'Swagger2->new->parse($yaml)->validate';
t 'Swagger2->new->parse($yaml)->expand->validate';

done_testing;

__DATA__
@@ swagger2.yaml
---
swagger: "2.0"
info:
  title: Example API
  version: "1.0"
basePath: /api
paths:
  /welcome:
    get:
      parameters:
      - name: param
        in: query
        required: true
        type: string
      responses:
        200:
          description: success
```

This is the output for me:

```plain
load.t ..
ok 1 - Swagger2->new($file)->validate
not ok 2 - Swagger2->new($file)->expand->validate

#   Failed test 'Swagger2->new($file)->expand->validate'
#   at load.t line 13.
#          got: '1'
#     expected: '0'
# [
#   bless( {
#     'message' => 'oneOf failed: ([0] oneOf failed: ([0] Properties not allowed: type. [1] oneOf failed: ([0] /in: Not in enum list: header. [2] /in: Not in enum list: formData. [7] /in: Not in enum list: path. [1] /required: Expected boolean - got number. [5] /required: Not in enum list: 1.)) [1] Properties not allowed: type, required, name, in.)',
#     'path' => '/paths/~1welcome/get/parameters/0'
#   }, 'JSON::Validator::Error' )
# ]
ok 3 - Swagger2->new($yaml)->validate
ok 4 - Swagger2->new($yaml)->expand->validate
ok 5 - Swagger2->new->load($file)->validate
ok 6 - Swagger2->new->load($file)->expand->validate
ok 7 - Swagger2->new->parse($yaml)->validate
ok 8 - Swagger2->new->parse($yaml)->expand->validate
1..8
# Looks like you failed 1 test of 8.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/8 subtests 

Test Summary Report
-------------------
load.t (Wstat: 256 Tests: 8 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
Files=1, Tests=8,  1 wallclock secs ( 0.02 usr  0.00 sys +  0.60 cusr  0.00 csys =  0.62 CPU)
Result: FAIL
```

I suggest:

* **not** marking `Swagger->new($url)` as deprecated because it's a reasonable, unambiguous, short, beloved syntax
* suggesting any of the alternatives as workaround
* fixing the bug and/or cleaning up Swagger2's internals to make all combinations work--I volunteer to help

**Comment by jhthorsen** [comment]
Just released 0.76, with the existing API working. Thanks for the test. It helped a lot.

There are two cases though which is not supported:
```
t 'Swagger2->new($yaml)->validate';
t 'Swagger2->new($yaml)->expand->validate';
```

For that to work I expect that new() would need to call parse() directly or I need some sort of ```$self->{text}``` which could be used inside the "api_spec" builder. I'm sorry to say that I'm not interested in either: #1: I don't like when new() does anything else than just building an object. #2: I think that "text" variable solution is fragile.
